nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/QueryIoTDBRecord.java

```diff
@@ -0,0 +1,230 @@
+            .name("record-writer")
+            .displayName("Record Writer")
+            .description("Specifies the Controller Service to use for writing results to a FlowFile. The Record Writer may use Inherit Schema to emulate the inferred schema behavior, i.e. "
+                    + "an explicit schema need not be defined in the writer, and will be supplied by the same logic used to infer the schema from the column types.")
+            .identifiesControllerService(RecordSetWriterFactory.class)
+            .required(true)
+            .build();
+
+    public static final PropertyDescriptor IOTDB_QUERY = new PropertyDescriptor.Builder()
+            .name("iotdb-query")
+            .displayName("IoTDB Query")
+            .description("The IoTDB query to execute. "
+                    + "Note: If there are incoming connections, then the query is created from incoming FlowFile's content otherwise"
+                    + " it is created from this property.")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .build();
+
+    public static final Integer FETCH_SIZE = 100000;
+    public static final PropertyDescriptor IOTDB_QUERY_FETCH_SIZE = new PropertyDescriptor.Builder()
+            .name("iotdb-query-chunk-size")
+            .displayName("Fetch Size")
+            .description("Chunking can be used to return results in a stream of smaller batches "
+                    + "(each has a partial results up to a chunk size) rather than as a single response. "
+                    + "Chunking queries can return an unlimited number of rows. Note: Chunking is enable when result chunk size is greater than 0")
+            .defaultValue(String.valueOf(DEFAULT_IOTDB_FETCH_SIZE))
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .addValidator(StandardValidators.createLongValidator(0, FETCH_SIZE, true))
+            .required(true)
+            .build();
+
+    private static RecordSetWriterFactory recordSetWriterFactory;
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> propertyDescriptors = new ArrayList<>(super.getSupportedPropertyDescriptors());
+        propertyDescriptors.add(IOTDB_QUERY);
+        propertyDescriptors.add(IOTDB_QUERY_FETCH_SIZE);
+        propertyDescriptors.add(RECORD_WRITER_FACTORY);
+        return Collections.unmodifiableList(propertyDescriptors);
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) throws IoTDBConnectionException {
+        super.onScheduled(context);
+        // Either input connection or scheduled query is required
+        if ( ! context.getProperty(IOTDB_QUERY).isSet()
+                && ! context.hasIncomingConnection() ) {
```

Comment:
Recommend running this file through automatic formatting to remove unnecessary spaces.
```suggestion
        if (!context.getProperty(IOTDB_QUERY).isSet()
                && ! context.hasIncomingConnection()) {
```